File: java/src/com/alofmethbin/jadawin/Article.java
package com.alofmethbin.jadawin;

import com.alofmethbin.jadawin.elements.Blurb;
import com.alofmethbin.jadawin.elements.Date;
import com.alofmethbin.jadawin.elements.Element;
import com.alofmethbin.jadawin.elements.Image;
import com.alofmethbin.jadawin.elements.Style;
import com.alofmethbin.jadawin.elements.Tag;
import com.alofmethbin.jadawin.elements.Title;
import com.alofmethbin.jadawin.styles.BaseStyle;
import com.alofmethbin.jadawin.styles.Story;
import java.io.File;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import com.alofmethbin.jadawin.styles.StyleInterface;
import java.io.IOException;

public final class Article implements Page {
    private List<Page> children = new ArrayList<>();
    private List<Element> content = new ArrayList<>();
    private Map<String,Element> specials = new HashMap<>();
    private StyleInterface style;
    private boolean childrenSorted = true;
    private final Article parent;
    private final String path;
    private Compiler compiler;
    private static StyleInterface baseStyle = new BaseStyle();
    
    class Origin {
        Origin( String n, Article a) {
            name    = n;
            article = a;
        }
        Article article;
        String name;
    }
    private List<Origin> origins = new ArrayList<>();
    
    public Article( Compiler compiler, Article parent, String path) {
        this.compiler = compiler;
        this.parent   = parent;
        this.path     = path;
    }
    
    public Page addChild( Page child) {
        childrenSorted = false;
        children.add( child);
        return child;
    }
    
    public void addContent( Element element, boolean multiline) {
        if (! element.hasError()) {
            if (multiline && (! element.isMultiline())) {
                error( element.typeName() + " takes only one line");
            } else if ( element.isSpecial() ) {
                if ( specials.containsKey( element.typeName())) {
                    error( "Duplicate " + element.typeName() + " definition");
                } else {
                    specials.put( element.typeName(), element);
                }
            } else {
                content.add( element);
            }
        }
    }

    public void addOrigin( String tag, Article parent) {
        origins.add( new Origin( tag, parent));
    }
    
    public List<Reference> breadcrumbs() {
        List<Reference> refs = new ArrayList<>();
        Article page = this;
        File file    = getSinkFile();
        
        while (page != null) {
            
            refs.add( 0, new Reference( Utils.relativePath( file, page.getSinkFile()), 
                                        Utils.prettify( page.getTitle())));
        }
        return refs;
    }
    
    public List<Page> children() {
        if (! this.childrenSorted) {
            getStyle().sort( children);
            childrenSorted = true;
        }
        return new ArrayList( children);
    }
    
    public void discardFutureChildren() {
        List<Page> old = children;
        children       = new ArrayList<>();
        
        for (Page child: old) {
            LocalDate d = child.getDate();
            if ((d == null) || (! d.isAfter( now()))) {
                children.add( child);
            }
        }
    }
    
    @Override
    public void error( String msg) {
        for (Element e: content) {
            if ( e.ignoreError( msg) ) {return;}
        }
        compiler.error( path, msg);
    }
    
    public String getBlurb() {
        Blurb blurb;
        if ((blurb = (Blurb) specials.get( "Blurb")) != null) {
            return blurb.text();
        }
        return null;
    }

    @Override
    public java.time.LocalDate getDate() {
        Date date;
        if ((date = (Date) specials.get( "Date")) != null) {
            return date.getDate();
        }
        return null;
    }
    
    @Override
    public Image getIcon() {
        Image icon;
        if ((icon = (Image) specials.get( "Icon")) != null) {
            return icon;
        }
        
        for (Element e: content) {
            if ((icon = e.getImage()) != null) {
                return icon;
            }
        }
        
        for (Page p: children()) {
            if ((icon = p.getIcon()) != null) {
                return icon;
            }
        }
        
        return null;
    }
    
    public String getName() {
        File file = getSinkFile();
        if ( file.getName().equals( "index.html") ) {
            return file.getParentFile().getName();
        } else {
            return file.getName().replace( ".html", "");
        }
    }
    
    public String getPageTitle() {
        if (parent != null) {
            return Utils.prettify( getTitle());
        } else {
            return null;
        }
    }

    public String getRootURL() {
        return compiler.getRootURL();
    }

    @Override
    public File getSinkFile() {
        return compiler.toSinkFile( path.replace( ".txt", ".html"));
    }
    
    public StyleInterface getStyle() {
        if (style != null) {
            return style;
        } else if ( isStyled() ) {
            return ((Style) specials.get( "Style")).getStyle();
        } else {
            return baseStyle;
        }
    }

    public List<Tag> getTags() {
        List<Tag> tags = new ArrayList<>();
        for (Element e: content) {
            if (e instanceof Tag) {
                tags.add( (Tag) e);
            }
        }
        return tags;
    }
    
    public String getTitle() {
        Title title;
        if ((title = (Title) specials.get( "Title")) != null) {
            return title.text();
        }
        
        return getName();
    }
    
    @Override
    public boolean hasChildren() {
        return ! children.isEmpty();
    }
    
    public boolean hasContent() {
        for (Element e: content) {
            if ( e.isContent() ) {return true;}
        }
        return false;
    }
    
    public boolean hasGrandChildren() {
        for (Page child: children) {
            if ( child.hasChildren() ) {return true;}
        }
        return false;
    }
    
    public boolean isLeaf() {
        return getStyle().isLeaf( this);
    }
    
    @Override
    public boolean isOffPage() {
        return false;
    }

    private boolean isStory() {
        return hasContent() && (getDate() != null) && (! hasGrandChildren());
    }

    public boolean isStyled() {
        return (specials.get( "Style") != null);
    }
    
    public boolean isWide() {
        for (Element e: content) {
            if ( e.isWide() ) {return true;}
        }
        return false;
    }

    public Page lookupPage( String url) {
        return compiler.lookupPage( this, url, url);
    }

    public LocalDate now() {
        return compiler.now();
    }

    private void overrideStyle( StyleInterface style) {
        this.style          = style;
        this.childrenSorted = false;
    }
    
    @Override
    public void prepare() {
        try {
            if ((! isStyled()) && isStory()) {
                overrideStyle( new Story());
            }
            getStyle().prepare( this);

            for (int i = 0; i < content.size(); i++) {
                content.get(i).prepare();
                if ( content.get(i).isInset() ) {
                    if (i < content.size() - 1) {
                        content.get(i+1).allowForInset();
                    } else {
                        error( "Inset cannot be final directive");
                    }
                }
            }
        } catch (Exception bang) {
            error( bang.getMessage());
            //throw bang;
        }
        
        List<Page> subArticles = new ArrayList<>();
        for (Page child: children) {
            if (child instanceof Article) {
                subArticles.add( child);
            }
        }
        
        for (Page child: subArticles) {
            child.prepare();
        }
    }

    public Article parent() {
        return parent;
    }
    
    public String postProcessHTML( String html) {
        html = getStyle().postProcessHTML( this, html);
        for (Element e: content) {
            html = e.postProcessHTML( html);
        }
        return html;
    }

    public void record( File sink) {
        compiler.record( sink);
    }

    @Override
    public void setStyle( StyleInterface style) {
        this.style = style;
    }

    public void spellCheck(String text) {
        compiler.spellCheck( this, text);
    }

    public String toPath( File file) throws IOException {
        return compiler.toPath( file);
    }

    public File toSinkFile( String relpath) {
        File file = getSinkFile();
        return new File( file.getParentFile(), relpath);
    }

    public File toSourceFile( String relpath) {
        File file = compiler.toSourceFile( path);
        return new File( file.getParentFile(), relpath);
    }
}

File: ruby/Article.rb
=begin
  Article.rb

  Represent article for HTML generation
=end

require 'digest/sha1'
require 'liquid'

require_relative 'utils'
require_relative 'elements/heading'
require_relative 'elements/icon'
require_relative 'elements/inset'
require_relative 'elements/links'
require_relative 'elements/tag'
require_relative 'elements/resource'
require_relative 'elements/svg'
require_relative 'styles/base'

class Article < Liquid::Drop
  include Utils
  attr_accessor :content_added
  attr_reader   :blurb, :style

  def initialize( compiler, parents, filename)
    @compiler        = compiler
    @parents         = parents
    @filename        = filename
    @children        = []
    @children_sorted = true
    @errors          = []
    @content         = []
    @specials        = {}
    @style           = nil
    # @style           = Styles::Base.new
    # @blurb           = nil
    # @date            = nil
    # @title           = nil
    @origins         = nil
#    @icon            = nil
  end

  def add_child( article)
    @children_sorted = false
    @children << article
  end

  def add_content( item, multiple_lines)
    unless item.discard?
      if multiple_lines && (! item.multiline?)
        error( item.type_name + ' takes only one line')
      elsif item.special?
        if @specials[ item.type_name]
          error( 'Duplicate ' + item.type_name + ' definition')
        else
          @specials[ item.type_name] = item
        end
      else
        @content << item
      end
    end
  end

  def blurb
    @specials['Blurb'] ? @specials['Blurb'].text : nil
  end

  def breadcrumbs( ancestors)
    return false if ancestors.empty?

    ancestors.collect do |ancestor|
      {'path'  => relative_path( filename, ancestor.filename),
       'title' => prettify( ancestor.title)}
    end
  end

  def children
    if not @children_sorted
      @children = style.sort( @children)
      @children_sorted = true
    end
    @children
  end

  def children?
    @children.size > 0
  end

  def date
    @specials['Date'] ? @specials['Date'].date : nil
  end

  def discard_future_children
    @children = @children.select do |child|
      child.date.nil? || (child.date <= Time.now)
    end
    self
  end

  def dump_content( tag)
    @content.each do |item|
      p [tag, item.class.to_s]
    end
  end

  def error( msg)
    @content.each do |element|
      return if element.ignore_error?( msg)
    end
    @errors << msg
  end

  def filename
    @filename
  end

  def grand_children?
    @children.each do |child|
      return true if child.children?
    end
    false
  end

  def has_any_content?
    # if /Baltics\/index/ =~ @filename
    #   puts 'DEBUG100'
    # end
    @content.each do |item|
      return true if item.page_content?
    end
    false
  end

  def icon
    return @specials['Icon'] if @specials['Icon']

    @content.each do |item|
      if icon = item.image
        return icon
      end
    end

    children.each do |child|
      if icon = child.icon
        return icon
      end
    end

    nil
  end

  def leaf?
    style.leaf?( self)
  end

  def name
    path = @filename.split( "/")
    (/^index\./i =~ path[-1]) ? path[-2] : path[-1].gsub( /\..*$/, '')
  end

  def off_page?
    false
  end

  def origin( name, ancestors)
    @origins = [] unless @origins
    @origins << [name, breadcrumbs( ancestors)]
  end

  def page_title
    @parents.empty? ? nil : prettify( title)
  end

  def post_process_html( root_url, html)
    html = style.post_process_html( root_url, self, html)
    @content.each do |element|
      html = element.post_process_html( self, html)
    end
    html
  end

  def prepare
    if (! styled?) && story?
      override_style( Styles::Story.new)
    end
    style.prepare( @compiler, self, @parents)

    @content.each_index do |i|
      @content[i].prepare( @compiler,
                           self,
                           @parents,
                           (@content.size > (i + 1)) ? @content[(i+1)..-1] : [])
    end
  end

  def report_errors
    @errors.each {|err| @compiler.error( @filename.gsub('html','txt'), err)}
  end

  def override_style( style)
    @style           = style
    @children_sorted = false
  end

  def story?
    has_any_content? && date && children? && (! grand_children?)
  end

  def style
    @style ? @style : (@specials['Style'] ? @specials['Style'].style : Styles::Base.new)
  end

  def styled?
    @specials['Style']
  end

  def tags
    @content.each do |element|
      yield element if element.is_a?( Elements::Tag)
    end
  end

  def text_index( articles, wide=true)
    index = 0
    articles.select {|a| a.style.index? && (wide || (! a.wide?))}.collect do |article|
      index += 1
      title = style.index_title(article)
      if title.size > 45
        while (title.size > 45) && (/ / =~ title)
          title = title.split( ' ')[0...-1].join( ' ')
        end
        title = title + ' ...'
      end
      {'path'     => relative_path( filename, article.filename),
       'off_page' => article.off_page?,
       'title'    => title.gsub( ' ', '&nbsp;'),
       'index'    => index
      }
    end
  end

  def title
    @specials['Title'] ? @specials['Title'].text : name
  end

  def to_data
    data = {}
    data['page_title']    = page_title
    data['blurb']         = blurb ? blurb : prettify( title)
    data['page_date']     = date ? format_date(date) : nil
    data['root']          = relative_path( filename, '/')
    data['breadcrumbs']   = breadcrumbs( @parents)
    data['origins']       = @origins

    if has_any_content?
      data['content']    = @content.select {|c| c.page_content?}.collect {|c| c.to_data( @compiler, self)}
      data['line_count'] = @content.inject(0) {|r,c| r + c.line_count}
      data['overlay']    = @content.inject(false) {|r,c| r || c.overlay?}
    end

    style.render( @compiler, @parents, self, data)
    data
  end

  def wide?
    @content.inject(false) {|r,e| r || e.wide?}
  end
end
